File: typescript/src/classifiers/anthropicClassifier.ts
import {
  ANTHROPIC_MODEL_ID_CLAUDE_3_5_SONNET,
  ConversationMessage,
  ParticipantRole,
} from "../types";
import { isClassifierToolInput } from "../utils/helpers";
import { Logger } from "../utils/logger";
import { Classifier, ClassifierResult } from "./classifier";
import { Anthropic } from "@anthropic-ai/sdk";

export interface AnthropicClassifierOptions {
  // Optional: The ID of the Anthropic model to use for classification
  // If not provided, a default model may be used
  modelId?: string;

  // Optional: Configuration for the inference process
  inferenceConfig?: {
    // Maximum number of tokens to generate in the response
    maxTokens?: number;

    // Controls randomness in output generation
    // Higher values (e.g., 0.8) make output more random, lower values (e.g., 0.2) make it more deterministic
    temperature?: number;

    // Controls diversity of output via nucleus sampling
    // 1.0 considers all tokens, lower values (e.g., 0.9) consider only the most probable tokens
    topP?: number;

    // Array of sequences that will stop the model from generating further tokens when encountered
    stopSequences?: string[];
  };

  // The API key for authenticating with Anthropic's services
  apiKey: string;
}

export class AnthropicClassifier extends Classifier {
  private client: Anthropic;
  protected inferenceConfig: {
    maxTokens?: number;
    temperature?: number;
    topP?: number;
    stopSequences?: string[];
  };

  private tools: Anthropic.Tool[] = [
    {
      name: 'analyzePrompt',
      description: 'Analyze the user input and provide structured output',
      input_schema: {
        type: 'object',
        properties: {
          userinput: {
            type: 'string',
            description: 'The original user input',
          },
          selected_agent: {
            type: 'string',
            description: 'The name of the selected agent',
          },
          confidence: {
            type: 'number',
            description: 'Confidence level between 0 and 1',
          },
        },
        required: ['userinput', 'selected_agent', 'confidence'],
      },
    },
  ];


  constructor(options: AnthropicClassifierOptions) {
    super();

    if (!options.apiKey) {
      throw new Error("Anthropic API key is required");
    }
    this.client = new Anthropic({ apiKey: options.apiKey });
    this.modelId = options.modelId || ANTHROPIC_MODEL_ID_CLAUDE_3_5_SONNET;
    // Set default value for max_tokens if not provided
    const defaultMaxTokens = 1000; // You can adjust this default value as needed
    this.inferenceConfig = {
      maxTokens: options.inferenceConfig?.maxTokens ?? defaultMaxTokens,
      temperature: options.inferenceConfig?.temperature,
      topP: options.inferenceConfig?.topP,
      stopSequences: options.inferenceConfig?.stopSequences,
    };

}

/* eslint-disable @typescript-eslint/no-unused-vars */
async processRequest(
    inputText: string,
    chatHistory: ConversationMessage[]
  ): Promise<ClassifierResult> {
    const userMessage: Anthropic.MessageParam = {
      role: ParticipantRole.USER,
      content: inputText,
    };

    try {
      const response = await this.client.messages.create({
        model: this.modelId,
        max_tokens: this.inferenceConfig.maxTokens,
        messages: [userMessage],
        system: this.systemPrompt,
        temperature: this.inferenceConfig.temperature,
        top_p: this.inferenceConfig.topP,
        tools: this.tools
      });

      const toolUse = response.content.find(
        (content): content is Anthropic.ToolUseBlock => content.type === "tool_use"
      );

      if (!toolUse) {
        throw new Error("No tool use found in the response");
      }

      if (!isClassifierToolInput(toolUse.input)) {
        throw new Error("Tool input does not match expected structure");
      }


      // Create and return IntentClassifierResult
      const intentClassifierResult: ClassifierResult = {
        selectedAgent: this.getAgentById(toolUse.input.selected_agent),
        confidence: parseFloat(toolUse.input.confidence),
      };
      return intentClassifierResult;

    } catch (error) {
      Logger.logger.error("Error processing request:", error);
      // Instead of returning a default result, we'll throw the error
      throw error;
    }
  }


}

File: examples/agentforce-demo/multi_agent_orchestrator/classifiers/anthropic_classifier.py
from typing import List, Optional, Dict, Any
from anthropic import Anthropic
from multi_agent_orchestrator.utils.helpers import is_tool_input
from multi_agent_orchestrator.utils.logger import Logger
from multi_agent_orchestrator.types import ConversationMessage
from multi_agent_orchestrator.classifiers import Classifier, ClassifierResult
import logging
logging.getLogger("httpx").setLevel(logging.WARNING)

ANTHROPIC_MODEL_ID_CLAUDE_3_5_SONNET = "claude-3-5-sonnet-20240620"

class AnthropicClassifierOptions:
    def __init__(self,
                 api_key: str,
                 model_id: Optional[str] = None,
                 inference_config: Optional[Dict[str, Any]] = None):
        self.api_key = api_key
        self.model_id = model_id
        self.inference_config = inference_config or {}

class AnthropicClassifier(Classifier):
    def __init__(self, options: AnthropicClassifierOptions):
        super().__init__()

        if not options.api_key:
            raise ValueError("Anthropic API key is required")

        self.client = Anthropic(api_key=options.api_key)
        self.model_id = options.model_id or ANTHROPIC_MODEL_ID_CLAUDE_3_5_SONNET

        default_max_tokens = 1000
        self.inference_config = {
            'max_tokens': options.inference_config.get('max_tokens', default_max_tokens),
            'temperature': options.inference_config.get('temperature', 0.0),
            'top_p': options.inference_config.get('top_p', 0.9),
            'stop_sequences': options.inference_config.get('stop_sequences', []),
        }

        self.tools: List[Dict] = [
            {
                'name': 'analyzePrompt',
                'description': 'Analyze the user input and provide structured output',
                'input_schema': {
                    'type': 'object',
                    'properties': {
                        'userinput': {
                            'type': 'string',
                            'description': 'The original user input',
                        },
                        'selected_agent': {
                            'type': 'string',
                            'description': 'The name of the selected agent',
                        },
                        'confidence': {
                            'type': 'number',
                            'description': 'Confidence level between 0 and 1',
                        },
                    },
                    'required': ['userinput', 'selected_agent', 'confidence'],
                },
            }
        ]

        self.system_prompt = "You are an AI assistant."  # Add your system prompt here


    async def process_request(self,
                              input_text: str,
                              chat_history: List[ConversationMessage]) -> ClassifierResult:
        user_message = {"role": "user", "content": input_text}

        try:
            response = self.client.messages.create(
                model=self.model_id,
                max_tokens=self.inference_config['max_tokens'],
                messages=[user_message],
                system=self.system_prompt,
                temperature=self.inference_config['temperature'],
                top_p=self.inference_config['top_p'],
                tools=self.tools
            )

            tool_use = next((c for c in response.content if c.type == "tool_use"), None)

            if not tool_use:
                raise ValueError("No tool use found in the response")

            if not is_tool_input(tool_use.input):
                raise ValueError("Tool input does not match expected structure")

            intent_classifier_result = ClassifierResult(
                selected_agent=self.get_agent_by_id(tool_use.input['selected_agent']),
                confidence=float(tool_use.input['confidence'])
            )

            return intent_classifier_result

        except Exception as error:
            Logger.error(f"Error processing request:{str(error)}")
            raise error
